{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "029ac793",
   "metadata": {},
   "outputs": [],
   "source": [
    "import polars as pl\n",
    "import bioframe as bf\n",
    "from IPython.display import display\n",
    "import os\n",
    "os.chdir(\"/zata/zippy/ramirezc/splice-model-benchmark\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4b044d0e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (2_742_017, 24)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>seqname</th><th>source</th><th>feature</th><th>start</th><th>end</th><th>score</th><th>strand</th><th>frame</th><th>gene_id</th><th>gene_type</th><th>gene_name</th><th>level</th><th>havana_gene</th><th>transcript_id</th><th>transcript_type</th><th>transcript_name</th><th>transcript_support_level</th><th>tag</th><th>havana_transcript</th><th>exon_number</th><th>exon_id</th><th>ont</th><th>protein_id</th><th>ccdsid</th></tr><tr><td>cat</td><td>cat</td><td>cat</td><td>i64</td><td>i64</td><td>f32</td><td>cat</td><td>i64</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td></tr></thead><tbody><tr><td>&quot;chr1&quot;</td><td>&quot;HAVANA&quot;</td><td>&quot;gene&quot;</td><td>11869</td><td>14409</td><td>null</td><td>&quot;+&quot;</td><td>0</td><td>&quot;ENSG00000223972.5&quot;</td><td>&quot;transcribed_unprocessed_pseudo…</td><td>&quot;DDX11L1&quot;</td><td>&quot;2&quot;</td><td>&quot;OTTHUMG00000000961.2&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td></tr><tr><td>&quot;chr1&quot;</td><td>&quot;HAVANA&quot;</td><td>&quot;transcript&quot;</td><td>11869</td><td>14409</td><td>null</td><td>&quot;+&quot;</td><td>0</td><td>&quot;ENSG00000223972.5&quot;</td><td>&quot;transcribed_unprocessed_pseudo…</td><td>&quot;DDX11L1&quot;</td><td>&quot;2&quot;</td><td>&quot;OTTHUMG00000000961.2&quot;</td><td>&quot;ENST00000456328.2&quot;</td><td>&quot;processed_transcript&quot;</td><td>&quot;DDX11L1-202&quot;</td><td>&quot;1&quot;</td><td>&quot;basic&quot;</td><td>&quot;OTTHUMT00000362751.1&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td></tr><tr><td>&quot;chr1&quot;</td><td>&quot;HAVANA&quot;</td><td>&quot;exon&quot;</td><td>11869</td><td>12227</td><td>null</td><td>&quot;+&quot;</td><td>0</td><td>&quot;ENSG00000223972.5&quot;</td><td>&quot;transcribed_unprocessed_pseudo…</td><td>&quot;DDX11L1&quot;</td><td>&quot;2&quot;</td><td>&quot;OTTHUMG00000000961.2&quot;</td><td>&quot;ENST00000456328.2&quot;</td><td>&quot;processed_transcript&quot;</td><td>&quot;DDX11L1-202&quot;</td><td>&quot;1&quot;</td><td>&quot;basic&quot;</td><td>&quot;OTTHUMT00000362751.1&quot;</td><td>&quot;1&quot;</td><td>&quot;ENSE00002234944.1&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td></tr><tr><td>&quot;chr1&quot;</td><td>&quot;HAVANA&quot;</td><td>&quot;exon&quot;</td><td>12613</td><td>12721</td><td>null</td><td>&quot;+&quot;</td><td>0</td><td>&quot;ENSG00000223972.5&quot;</td><td>&quot;transcribed_unprocessed_pseudo…</td><td>&quot;DDX11L1&quot;</td><td>&quot;2&quot;</td><td>&quot;OTTHUMG00000000961.2&quot;</td><td>&quot;ENST00000456328.2&quot;</td><td>&quot;processed_transcript&quot;</td><td>&quot;DDX11L1-202&quot;</td><td>&quot;1&quot;</td><td>&quot;basic&quot;</td><td>&quot;OTTHUMT00000362751.1&quot;</td><td>&quot;2&quot;</td><td>&quot;ENSE00003582793.1&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td></tr><tr><td>&quot;chr1&quot;</td><td>&quot;HAVANA&quot;</td><td>&quot;exon&quot;</td><td>13221</td><td>14409</td><td>null</td><td>&quot;+&quot;</td><td>0</td><td>&quot;ENSG00000223972.5&quot;</td><td>&quot;transcribed_unprocessed_pseudo…</td><td>&quot;DDX11L1&quot;</td><td>&quot;2&quot;</td><td>&quot;OTTHUMG00000000961.2&quot;</td><td>&quot;ENST00000456328.2&quot;</td><td>&quot;processed_transcript&quot;</td><td>&quot;DDX11L1-202&quot;</td><td>&quot;1&quot;</td><td>&quot;basic&quot;</td><td>&quot;OTTHUMT00000362751.1&quot;</td><td>&quot;3&quot;</td><td>&quot;ENSE00002312635.1&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;chrM&quot;</td><td>&quot;ENSEMBL&quot;</td><td>&quot;transcript&quot;</td><td>15888</td><td>15953</td><td>null</td><td>&quot;+&quot;</td><td>0</td><td>&quot;ENSG00000210195.2&quot;</td><td>&quot;Mt_tRNA&quot;</td><td>&quot;MT-TT&quot;</td><td>&quot;3&quot;</td><td>&quot;&quot;</td><td>&quot;ENST00000387460.2&quot;</td><td>&quot;Mt_tRNA&quot;</td><td>&quot;MT-TT-201&quot;</td><td>&quot;NA&quot;</td><td>&quot;basic&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td></tr><tr><td>&quot;chrM&quot;</td><td>&quot;ENSEMBL&quot;</td><td>&quot;exon&quot;</td><td>15888</td><td>15953</td><td>null</td><td>&quot;+&quot;</td><td>0</td><td>&quot;ENSG00000210195.2&quot;</td><td>&quot;Mt_tRNA&quot;</td><td>&quot;MT-TT&quot;</td><td>&quot;3&quot;</td><td>&quot;&quot;</td><td>&quot;ENST00000387460.2&quot;</td><td>&quot;Mt_tRNA&quot;</td><td>&quot;MT-TT-201&quot;</td><td>&quot;NA&quot;</td><td>&quot;basic&quot;</td><td>&quot;&quot;</td><td>&quot;1&quot;</td><td>&quot;ENSE00001544475.2&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td></tr><tr><td>&quot;chrM&quot;</td><td>&quot;ENSEMBL&quot;</td><td>&quot;gene&quot;</td><td>15956</td><td>16023</td><td>null</td><td>&quot;-&quot;</td><td>0</td><td>&quot;ENSG00000210196.2&quot;</td><td>&quot;Mt_tRNA&quot;</td><td>&quot;MT-TP&quot;</td><td>&quot;3&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td></tr><tr><td>&quot;chrM&quot;</td><td>&quot;ENSEMBL&quot;</td><td>&quot;transcript&quot;</td><td>15956</td><td>16023</td><td>null</td><td>&quot;-&quot;</td><td>0</td><td>&quot;ENSG00000210196.2&quot;</td><td>&quot;Mt_tRNA&quot;</td><td>&quot;MT-TP&quot;</td><td>&quot;3&quot;</td><td>&quot;&quot;</td><td>&quot;ENST00000387461.2&quot;</td><td>&quot;Mt_tRNA&quot;</td><td>&quot;MT-TP-201&quot;</td><td>&quot;NA&quot;</td><td>&quot;basic&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td></tr><tr><td>&quot;chrM&quot;</td><td>&quot;ENSEMBL&quot;</td><td>&quot;exon&quot;</td><td>15956</td><td>16023</td><td>null</td><td>&quot;-&quot;</td><td>0</td><td>&quot;ENSG00000210196.2&quot;</td><td>&quot;Mt_tRNA&quot;</td><td>&quot;MT-TP&quot;</td><td>&quot;3&quot;</td><td>&quot;&quot;</td><td>&quot;ENST00000387461.2&quot;</td><td>&quot;Mt_tRNA&quot;</td><td>&quot;MT-TP-201&quot;</td><td>&quot;NA&quot;</td><td>&quot;basic&quot;</td><td>&quot;&quot;</td><td>&quot;1&quot;</td><td>&quot;ENSE00001544473.2&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td><td>&quot;&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (2_742_017, 24)\n",
       "┌─────────┬─────────┬────────────┬───────┬───┬───────────────────┬─────┬────────────┬────────┐\n",
       "│ seqname ┆ source  ┆ feature    ┆ start ┆ … ┆ exon_id           ┆ ont ┆ protein_id ┆ ccdsid │\n",
       "│ ---     ┆ ---     ┆ ---        ┆ ---   ┆   ┆ ---               ┆ --- ┆ ---        ┆ ---    │\n",
       "│ cat     ┆ cat     ┆ cat        ┆ i64   ┆   ┆ str               ┆ str ┆ str        ┆ str    │\n",
       "╞═════════╪═════════╪════════════╪═══════╪═══╪═══════════════════╪═════╪════════════╪════════╡\n",
       "│ chr1    ┆ HAVANA  ┆ gene       ┆ 11869 ┆ … ┆                   ┆     ┆            ┆        │\n",
       "│ chr1    ┆ HAVANA  ┆ transcript ┆ 11869 ┆ … ┆                   ┆     ┆            ┆        │\n",
       "│ chr1    ┆ HAVANA  ┆ exon       ┆ 11869 ┆ … ┆ ENSE00002234944.1 ┆     ┆            ┆        │\n",
       "│ chr1    ┆ HAVANA  ┆ exon       ┆ 12613 ┆ … ┆ ENSE00003582793.1 ┆     ┆            ┆        │\n",
       "│ chr1    ┆ HAVANA  ┆ exon       ┆ 13221 ┆ … ┆ ENSE00002312635.1 ┆     ┆            ┆        │\n",
       "│ …       ┆ …       ┆ …          ┆ …     ┆ … ┆ …                 ┆ …   ┆ …          ┆ …      │\n",
       "│ chrM    ┆ ENSEMBL ┆ transcript ┆ 15888 ┆ … ┆                   ┆     ┆            ┆        │\n",
       "│ chrM    ┆ ENSEMBL ┆ exon       ┆ 15888 ┆ … ┆ ENSE00001544475.2 ┆     ┆            ┆        │\n",
       "│ chrM    ┆ ENSEMBL ┆ gene       ┆ 15956 ┆ … ┆                   ┆     ┆            ┆        │\n",
       "│ chrM    ┆ ENSEMBL ┆ transcript ┆ 15956 ┆ … ┆                   ┆     ┆            ┆        │\n",
       "│ chrM    ┆ ENSEMBL ┆ exon       ┆ 15956 ┆ … ┆ ENSE00001544473.2 ┆     ┆            ┆        │\n",
       "└─────────┴─────────┴────────────┴───────┴───┴───────────────────┴─────┴────────────┴────────┘"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (222_782, 6)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>chrom</th><th>start</th><th>end</th><th>strand</th><th>reads</th><th>reads_per_million</th></tr><tr><td>str</td><td>i64</td><td>i64</td><td>str</td><td>i64</td><td>f64</td></tr></thead><tbody><tr><td>&quot;chr1&quot;</td><td>827775</td><td>829002</td><td>&quot;+&quot;</td><td>4</td><td>2.365377</td></tr><tr><td>&quot;chr1&quot;</td><td>829104</td><td>841199</td><td>&quot;+&quot;</td><td>6</td><td>3.548066</td></tr><tr><td>&quot;chr1&quot;</td><td>827853</td><td>829002</td><td>&quot;+&quot;</td><td>1</td><td>0.591344</td></tr><tr><td>&quot;chr1&quot;</td><td>960800</td><td>961628</td><td>&quot;+&quot;</td><td>5</td><td>2.956722</td></tr><tr><td>&quot;chr1&quot;</td><td>961750</td><td>961825</td><td>&quot;+&quot;</td><td>16</td><td>9.46151</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;chrX&quot;</td><td>155514265</td><td>155524455</td><td>&quot;-&quot;</td><td>11</td><td>6.504788</td></tr><tr><td>&quot;chrX&quot;</td><td>155524632</td><td>155612791</td><td>&quot;-&quot;</td><td>2</td><td>1.182689</td></tr><tr><td>&quot;chrX&quot;</td><td>155524632</td><td>155545095</td><td>&quot;-&quot;</td><td>12</td><td>7.096132</td></tr><tr><td>&quot;chrX&quot;</td><td>155545277</td><td>155612791</td><td>&quot;-&quot;</td><td>12</td><td>7.096132</td></tr><tr><td>&quot;chrX&quot;</td><td>155492700</td><td>155506897</td><td>&quot;-&quot;</td><td>2</td><td>1.182689</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (222_782, 6)\n",
       "┌───────┬───────────┬───────────┬────────┬───────┬───────────────────┐\n",
       "│ chrom ┆ start     ┆ end       ┆ strand ┆ reads ┆ reads_per_million │\n",
       "│ ---   ┆ ---       ┆ ---       ┆ ---    ┆ ---   ┆ ---               │\n",
       "│ str   ┆ i64       ┆ i64       ┆ str    ┆ i64   ┆ f64               │\n",
       "╞═══════╪═══════════╪═══════════╪════════╪═══════╪═══════════════════╡\n",
       "│ chr1  ┆ 827775    ┆ 829002    ┆ +      ┆ 4     ┆ 2.365377          │\n",
       "│ chr1  ┆ 829104    ┆ 841199    ┆ +      ┆ 6     ┆ 3.548066          │\n",
       "│ chr1  ┆ 827853    ┆ 829002    ┆ +      ┆ 1     ┆ 0.591344          │\n",
       "│ chr1  ┆ 960800    ┆ 961628    ┆ +      ┆ 5     ┆ 2.956722          │\n",
       "│ chr1  ┆ 961750    ┆ 961825    ┆ +      ┆ 16    ┆ 9.46151           │\n",
       "│ …     ┆ …         ┆ …         ┆ …      ┆ …     ┆ …                 │\n",
       "│ chrX  ┆ 155514265 ┆ 155524455 ┆ -      ┆ 11    ┆ 6.504788          │\n",
       "│ chrX  ┆ 155524632 ┆ 155612791 ┆ -      ┆ 2     ┆ 1.182689          │\n",
       "│ chrX  ┆ 155524632 ┆ 155545095 ┆ -      ┆ 12    ┆ 7.096132          │\n",
       "│ chrX  ┆ 155545277 ┆ 155612791 ┆ -      ┆ 12    ┆ 7.096132          │\n",
       "│ chrX  ┆ 155492700 ┆ 155506897 ┆ -      ┆ 2     ┆ 1.182689          │\n",
       "└───────┴───────────┴───────────┴────────┴───────┴───────────────────┘"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gtf = pl.read_parquet('/zata/zippy/ramirezc/static_files/gencode.v29.basic.annotation.gtf.parquet')\n",
    "display(gtf)\n",
    "example_intron_df = pl.read_csv('golden_standard/downloads/ENCBS020DQC/ENCFF835KGV.introns.tab', separator='\\t')\n",
    "display(example_intron_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "98bb8fd3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (19_940, 9)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>chrom</th><th>start</th><th>end</th><th>strand</th><th>feature</th><th>gene_id</th><th>gene_name</th><th>transcript_id</th><th>gene_type</th></tr><tr><td>cat</td><td>i64</td><td>i64</td><td>cat</td><td>cat</td><td>str</td><td>str</td><td>str</td><td>str</td></tr></thead><tbody><tr><td>&quot;chr1&quot;</td><td>65419</td><td>71585</td><td>&quot;+&quot;</td><td>&quot;gene&quot;</td><td>&quot;ENSG00000186092.6&quot;</td><td>&quot;OR4F5&quot;</td><td>&quot;&quot;</td><td>&quot;protein_coding&quot;</td></tr><tr><td>&quot;chr1&quot;</td><td>450703</td><td>451697</td><td>&quot;-&quot;</td><td>&quot;gene&quot;</td><td>&quot;ENSG00000284733.1&quot;</td><td>&quot;OR4F29&quot;</td><td>&quot;&quot;</td><td>&quot;protein_coding&quot;</td></tr><tr><td>&quot;chr1&quot;</td><td>685679</td><td>686673</td><td>&quot;-&quot;</td><td>&quot;gene&quot;</td><td>&quot;ENSG00000284662.1&quot;</td><td>&quot;OR4F16&quot;</td><td>&quot;&quot;</td><td>&quot;protein_coding&quot;</td></tr><tr><td>&quot;chr1&quot;</td><td>923928</td><td>944581</td><td>&quot;+&quot;</td><td>&quot;gene&quot;</td><td>&quot;ENSG00000187634.11&quot;</td><td>&quot;SAMD11&quot;</td><td>&quot;&quot;</td><td>&quot;protein_coding&quot;</td></tr><tr><td>&quot;chr1&quot;</td><td>944204</td><td>959309</td><td>&quot;-&quot;</td><td>&quot;gene&quot;</td><td>&quot;ENSG00000188976.10&quot;</td><td>&quot;NOC2L&quot;</td><td>&quot;&quot;</td><td>&quot;protein_coding&quot;</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;chrM&quot;</td><td>10470</td><td>10766</td><td>&quot;+&quot;</td><td>&quot;gene&quot;</td><td>&quot;ENSG00000212907.2&quot;</td><td>&quot;MT-ND4L&quot;</td><td>&quot;&quot;</td><td>&quot;protein_coding&quot;</td></tr><tr><td>&quot;chrM&quot;</td><td>10760</td><td>12137</td><td>&quot;+&quot;</td><td>&quot;gene&quot;</td><td>&quot;ENSG00000198886.2&quot;</td><td>&quot;MT-ND4&quot;</td><td>&quot;&quot;</td><td>&quot;protein_coding&quot;</td></tr><tr><td>&quot;chrM&quot;</td><td>12337</td><td>14148</td><td>&quot;+&quot;</td><td>&quot;gene&quot;</td><td>&quot;ENSG00000198786.2&quot;</td><td>&quot;MT-ND5&quot;</td><td>&quot;&quot;</td><td>&quot;protein_coding&quot;</td></tr><tr><td>&quot;chrM&quot;</td><td>14149</td><td>14673</td><td>&quot;-&quot;</td><td>&quot;gene&quot;</td><td>&quot;ENSG00000198695.2&quot;</td><td>&quot;MT-ND6&quot;</td><td>&quot;&quot;</td><td>&quot;protein_coding&quot;</td></tr><tr><td>&quot;chrM&quot;</td><td>14747</td><td>15887</td><td>&quot;+&quot;</td><td>&quot;gene&quot;</td><td>&quot;ENSG00000198727.2&quot;</td><td>&quot;MT-CYB&quot;</td><td>&quot;&quot;</td><td>&quot;protein_coding&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (19_940, 9)\n",
       "┌───────┬────────┬────────┬────────┬───┬───────────────┬───────────┬───────────────┬───────────────┐\n",
       "│ chrom ┆ start  ┆ end    ┆ strand ┆ … ┆ gene_id       ┆ gene_name ┆ transcript_id ┆ gene_type     │\n",
       "│ ---   ┆ ---    ┆ ---    ┆ ---    ┆   ┆ ---           ┆ ---       ┆ ---           ┆ ---           │\n",
       "│ cat   ┆ i64    ┆ i64    ┆ cat    ┆   ┆ str           ┆ str       ┆ str           ┆ str           │\n",
       "╞═══════╪════════╪════════╪════════╪═══╪═══════════════╪═══════════╪═══════════════╪═══════════════╡\n",
       "│ chr1  ┆ 65419  ┆ 71585  ┆ +      ┆ … ┆ ENSG000001860 ┆ OR4F5     ┆               ┆ protein_codin │\n",
       "│       ┆        ┆        ┆        ┆   ┆ 92.6          ┆           ┆               ┆ g             │\n",
       "│ chr1  ┆ 450703 ┆ 451697 ┆ -      ┆ … ┆ ENSG000002847 ┆ OR4F29    ┆               ┆ protein_codin │\n",
       "│       ┆        ┆        ┆        ┆   ┆ 33.1          ┆           ┆               ┆ g             │\n",
       "│ chr1  ┆ 685679 ┆ 686673 ┆ -      ┆ … ┆ ENSG000002846 ┆ OR4F16    ┆               ┆ protein_codin │\n",
       "│       ┆        ┆        ┆        ┆   ┆ 62.1          ┆           ┆               ┆ g             │\n",
       "│ chr1  ┆ 923928 ┆ 944581 ┆ +      ┆ … ┆ ENSG000001876 ┆ SAMD11    ┆               ┆ protein_codin │\n",
       "│       ┆        ┆        ┆        ┆   ┆ 34.11         ┆           ┆               ┆ g             │\n",
       "│ chr1  ┆ 944204 ┆ 959309 ┆ -      ┆ … ┆ ENSG000001889 ┆ NOC2L     ┆               ┆ protein_codin │\n",
       "│       ┆        ┆        ┆        ┆   ┆ 76.10         ┆           ┆               ┆ g             │\n",
       "│ …     ┆ …      ┆ …      ┆ …      ┆ … ┆ …             ┆ …         ┆ …             ┆ …             │\n",
       "│ chrM  ┆ 10470  ┆ 10766  ┆ +      ┆ … ┆ ENSG000002129 ┆ MT-ND4L   ┆               ┆ protein_codin │\n",
       "│       ┆        ┆        ┆        ┆   ┆ 07.2          ┆           ┆               ┆ g             │\n",
       "│ chrM  ┆ 10760  ┆ 12137  ┆ +      ┆ … ┆ ENSG000001988 ┆ MT-ND4    ┆               ┆ protein_codin │\n",
       "│       ┆        ┆        ┆        ┆   ┆ 86.2          ┆           ┆               ┆ g             │\n",
       "│ chrM  ┆ 12337  ┆ 14148  ┆ +      ┆ … ┆ ENSG000001987 ┆ MT-ND5    ┆               ┆ protein_codin │\n",
       "│       ┆        ┆        ┆        ┆   ┆ 86.2          ┆           ┆               ┆ g             │\n",
       "│ chrM  ┆ 14149  ┆ 14673  ┆ -      ┆ … ┆ ENSG000001986 ┆ MT-ND6    ┆               ┆ protein_codin │\n",
       "│       ┆        ┆        ┆        ┆   ┆ 95.2          ┆           ┆               ┆ g             │\n",
       "│ chrM  ┆ 14747  ┆ 15887  ┆ +      ┆ … ┆ ENSG000001987 ┆ MT-CYB    ┆               ┆ protein_codin │\n",
       "│       ┆        ┆        ┆        ┆   ┆ 27.2          ┆           ┆               ┆ g             │\n",
       "└───────┴────────┴────────┴────────┴───┴───────────────┴───────────┴───────────────┴───────────────┘"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "slim_gtf = gtf.select(['seqname', 'start', 'end', 'strand', 'feature', 'gene_id', 'gene_name', 'transcript_id', 'gene_type']).rename({'seqname': 'chrom'})\n",
    "filter_gtf = slim_gtf.filter(pl.col('feature') == 'gene', pl.col('gene_type') == 'protein_coding')\n",
    "display(filter_gtf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0441d30e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom_gene</th>\n",
       "      <th>start_gene</th>\n",
       "      <th>end_gene</th>\n",
       "      <th>strand_gene</th>\n",
       "      <th>feature_gene</th>\n",
       "      <th>gene_id_gene</th>\n",
       "      <th>gene_name_gene</th>\n",
       "      <th>transcript_id_gene</th>\n",
       "      <th>gene_type_gene</th>\n",
       "      <th>chrom_intron</th>\n",
       "      <th>start_intron</th>\n",
       "      <th>end_intron</th>\n",
       "      <th>strand_intron</th>\n",
       "      <th>reads_intron</th>\n",
       "      <th>reads_per_million_intron</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr7</td>\n",
       "      <td>192969</td>\n",
       "      <td>260745</td>\n",
       "      <td>+</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSG00000177706.8</td>\n",
       "      <td>FAM20C</td>\n",
       "      <td></td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr7</td>\n",
       "      <td>193804</td>\n",
       "      <td>195553</td>\n",
       "      <td>+</td>\n",
       "      <td>13</td>\n",
       "      <td>7.687477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr7</td>\n",
       "      <td>192969</td>\n",
       "      <td>260745</td>\n",
       "      <td>+</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSG00000177706.8</td>\n",
       "      <td>FAM20C</td>\n",
       "      <td></td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr7</td>\n",
       "      <td>193804</td>\n",
       "      <td>208897</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>0.591344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr7</td>\n",
       "      <td>192969</td>\n",
       "      <td>260745</td>\n",
       "      <td>+</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSG00000177706.8</td>\n",
       "      <td>FAM20C</td>\n",
       "      <td></td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr7</td>\n",
       "      <td>195732</td>\n",
       "      <td>208897</td>\n",
       "      <td>+</td>\n",
       "      <td>12</td>\n",
       "      <td>7.096132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr7</td>\n",
       "      <td>192969</td>\n",
       "      <td>260745</td>\n",
       "      <td>+</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSG00000177706.8</td>\n",
       "      <td>FAM20C</td>\n",
       "      <td></td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr7</td>\n",
       "      <td>208976</td>\n",
       "      <td>246414</td>\n",
       "      <td>+</td>\n",
       "      <td>13</td>\n",
       "      <td>7.687477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr7</td>\n",
       "      <td>192969</td>\n",
       "      <td>260745</td>\n",
       "      <td>+</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSG00000177706.8</td>\n",
       "      <td>FAM20C</td>\n",
       "      <td></td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr7</td>\n",
       "      <td>246507</td>\n",
       "      <td>248314</td>\n",
       "      <td>+</td>\n",
       "      <td>13</td>\n",
       "      <td>7.687477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236492</th>\n",
       "      <td>chr18</td>\n",
       "      <td>49488453</td>\n",
       "      <td>49492523</td>\n",
       "      <td>-</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSG00000265681.7</td>\n",
       "      <td>RPL17</td>\n",
       "      <td></td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr18</td>\n",
       "      <td>49483771</td>\n",
       "      <td>49489358</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>1.182689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236493</th>\n",
       "      <td>chr18</td>\n",
       "      <td>49795793</td>\n",
       "      <td>49837623</td>\n",
       "      <td>-</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSG00000266997.2</td>\n",
       "      <td>AC090227.1</td>\n",
       "      <td></td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr18</td>\n",
       "      <td>49787361</td>\n",
       "      <td>49813816</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>0.591344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236494</th>\n",
       "      <td>chr18</td>\n",
       "      <td>49795793</td>\n",
       "      <td>49837623</td>\n",
       "      <td>-</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSG00000266997.2</td>\n",
       "      <td>AC090227.1</td>\n",
       "      <td></td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr18</td>\n",
       "      <td>49794427</td>\n",
       "      <td>49797465</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>0.591344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236495</th>\n",
       "      <td>chr18</td>\n",
       "      <td>57646426</td>\n",
       "      <td>57803315</td>\n",
       "      <td>-</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSG00000081923.13</td>\n",
       "      <td>ATP8B1</td>\n",
       "      <td></td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr18</td>\n",
       "      <td>57601783</td>\n",
       "      <td>57655193</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "      <td>0.591344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>236496</th>\n",
       "      <td>chr18</td>\n",
       "      <td>57646426</td>\n",
       "      <td>57803315</td>\n",
       "      <td>-</td>\n",
       "      <td>gene</td>\n",
       "      <td>ENSG00000081923.13</td>\n",
       "      <td>ATP8B1</td>\n",
       "      <td></td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr18</td>\n",
       "      <td>57642052</td>\n",
       "      <td>57668005</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>0.591344</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>236497 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chrom_gene  start_gene  end_gene strand_gene feature_gene  \\\n",
       "0            chr7      192969    260745           +         gene   \n",
       "1            chr7      192969    260745           +         gene   \n",
       "2            chr7      192969    260745           +         gene   \n",
       "3            chr7      192969    260745           +         gene   \n",
       "4            chr7      192969    260745           +         gene   \n",
       "...           ...         ...       ...         ...          ...   \n",
       "236492      chr18    49488453  49492523           -         gene   \n",
       "236493      chr18    49795793  49837623           -         gene   \n",
       "236494      chr18    49795793  49837623           -         gene   \n",
       "236495      chr18    57646426  57803315           -         gene   \n",
       "236496      chr18    57646426  57803315           -         gene   \n",
       "\n",
       "              gene_id_gene gene_name_gene transcript_id_gene  gene_type_gene  \\\n",
       "0        ENSG00000177706.8         FAM20C                     protein_coding   \n",
       "1        ENSG00000177706.8         FAM20C                     protein_coding   \n",
       "2        ENSG00000177706.8         FAM20C                     protein_coding   \n",
       "3        ENSG00000177706.8         FAM20C                     protein_coding   \n",
       "4        ENSG00000177706.8         FAM20C                     protein_coding   \n",
       "...                    ...            ...                ...             ...   \n",
       "236492   ENSG00000265681.7          RPL17                     protein_coding   \n",
       "236493   ENSG00000266997.2     AC090227.1                     protein_coding   \n",
       "236494   ENSG00000266997.2     AC090227.1                     protein_coding   \n",
       "236495  ENSG00000081923.13         ATP8B1                     protein_coding   \n",
       "236496  ENSG00000081923.13         ATP8B1                     protein_coding   \n",
       "\n",
       "       chrom_intron  start_intron  end_intron strand_intron  reads_intron  \\\n",
       "0              chr7        193804      195553             +            13   \n",
       "1              chr7        193804      208897             +             1   \n",
       "2              chr7        195732      208897             +            12   \n",
       "3              chr7        208976      246414             +            13   \n",
       "4              chr7        246507      248314             +            13   \n",
       "...             ...           ...         ...           ...           ...   \n",
       "236492        chr18      49483771    49489358             -             2   \n",
       "236493        chr18      49787361    49813816             -             1   \n",
       "236494        chr18      49794427    49797465             -             1   \n",
       "236495        chr18      57601783    57655193             -             1   \n",
       "236496        chr18      57642052    57668005             +             1   \n",
       "\n",
       "        reads_per_million_intron  \n",
       "0                       7.687477  \n",
       "1                       0.591344  \n",
       "2                       7.096132  \n",
       "3                       7.687477  \n",
       "4                       7.687477  \n",
       "...                          ...  \n",
       "236492                  1.182689  \n",
       "236493                  0.591344  \n",
       "236494                  0.591344  \n",
       "236495                  0.591344  \n",
       "236496                  0.591344  \n",
       "\n",
       "[236497 rows x 15 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (211_178, 6)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>chrom</th><th>start</th><th>end</th><th>strand</th><th>reads</th><th>gene_name</th></tr><tr><td>str</td><td>i64</td><td>i64</td><td>str</td><td>i64</td><td>str</td></tr></thead><tbody><tr><td>&quot;chr1&quot;</td><td>944550</td><td>945041</td><td>&quot;-&quot;</td><td>1</td><td>&quot;NOC2L&quot;</td></tr><tr><td>&quot;chr1&quot;</td><td>944750</td><td>945056</td><td>&quot;-&quot;</td><td>1</td><td>&quot;NOC2L&quot;</td></tr><tr><td>&quot;chr1&quot;</td><td>944800</td><td>945056</td><td>&quot;-&quot;</td><td>284</td><td>&quot;NOC2L&quot;</td></tr><tr><td>&quot;chr1&quot;</td><td>944800</td><td>945041</td><td>&quot;-&quot;</td><td>141</td><td>&quot;NOC2L&quot;</td></tr><tr><td>&quot;chr1&quot;</td><td>944819</td><td>945041</td><td>&quot;-&quot;</td><td>1</td><td>&quot;NOC2L&quot;</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;chrX&quot;</td><td>155898249</td><td>155919812</td><td>&quot;+&quot;</td><td>1</td><td>&quot;VAMP7&quot;</td></tr><tr><td>&quot;chrX&quot;</td><td>155900587</td><td>155939700</td><td>&quot;+&quot;</td><td>2</td><td>&quot;VAMP7&quot;</td></tr><tr><td>&quot;chrX&quot;</td><td>155900587</td><td>155919812</td><td>&quot;+&quot;</td><td>83</td><td>&quot;VAMP7&quot;</td></tr><tr><td>&quot;chrX&quot;</td><td>155919880</td><td>155939700</td><td>&quot;+&quot;</td><td>84</td><td>&quot;VAMP7&quot;</td></tr><tr><td>&quot;chrX&quot;</td><td>155939793</td><td>155941882</td><td>&quot;+&quot;</td><td>86</td><td>&quot;VAMP7&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (211_178, 6)\n",
       "┌───────┬───────────┬───────────┬────────┬───────┬───────────┐\n",
       "│ chrom ┆ start     ┆ end       ┆ strand ┆ reads ┆ gene_name │\n",
       "│ ---   ┆ ---       ┆ ---       ┆ ---    ┆ ---   ┆ ---       │\n",
       "│ str   ┆ i64       ┆ i64       ┆ str    ┆ i64   ┆ str       │\n",
       "╞═══════╪═══════════╪═══════════╪════════╪═══════╪═══════════╡\n",
       "│ chr1  ┆ 944550    ┆ 945041    ┆ -      ┆ 1     ┆ NOC2L     │\n",
       "│ chr1  ┆ 944750    ┆ 945056    ┆ -      ┆ 1     ┆ NOC2L     │\n",
       "│ chr1  ┆ 944800    ┆ 945056    ┆ -      ┆ 284   ┆ NOC2L     │\n",
       "│ chr1  ┆ 944800    ┆ 945041    ┆ -      ┆ 141   ┆ NOC2L     │\n",
       "│ chr1  ┆ 944819    ┆ 945041    ┆ -      ┆ 1     ┆ NOC2L     │\n",
       "│ …     ┆ …         ┆ …         ┆ …      ┆ …     ┆ …         │\n",
       "│ chrX  ┆ 155898249 ┆ 155919812 ┆ +      ┆ 1     ┆ VAMP7     │\n",
       "│ chrX  ┆ 155900587 ┆ 155939700 ┆ +      ┆ 2     ┆ VAMP7     │\n",
       "│ chrX  ┆ 155900587 ┆ 155919812 ┆ +      ┆ 83    ┆ VAMP7     │\n",
       "│ chrX  ┆ 155919880 ┆ 155939700 ┆ +      ┆ 84    ┆ VAMP7     │\n",
       "│ chrX  ┆ 155939793 ┆ 155941882 ┆ +      ┆ 86    ┆ VAMP7     │\n",
       "└───────┴───────────┴───────────┴────────┴───────┴───────────┘"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "overlaps = bf.overlap(\n",
    "    filter_gtf.to_pandas(),\n",
    "    example_intron_df.to_pandas(),\n",
    "    how='inner',\n",
    "    suffixes=('_gene', '_intron')\n",
    ")\n",
    "display(overlaps)\n",
    "\n",
    "contained = overlaps[\n",
    "    (overlaps['start_intron'] >= overlaps['start_gene']) & \n",
    "    (overlaps['end_intron'] <= overlaps['end_gene'])\n",
    "]\n",
    "contained_pl = pl.from_pandas(contained)\n",
    "unique_introns = contained_pl.unique(\n",
    "    subset=['chrom_intron', 'start_intron', 'end_intron', 'strand_intron', 'reads_intron']\n",
    ").rename(\n",
    "    {\n",
    "        'chrom_intron': 'chrom',\n",
    "        'start_intron': 'start',\n",
    "        'end_intron': 'end',\n",
    "        'strand_intron': 'strand',\n",
    "        'reads_intron': 'reads',\n",
    "        'gene_name_gene': 'gene_name',\n",
    "    }\n",
    ").select('chrom', 'start', 'end', 'strand', 'reads', 'gene_name').sort('chrom', 'start')\n",
    "\n",
    "display(unique_introns)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ee5af02",
   "metadata": {},
   "source": [
    "Started with 225_467 introns\n",
    "\n",
    "~~After filtering, there are 230_405 (buh)~~\n",
    "\n",
    "Had to filter for unique introns, because the gtf file has some overlapping genes don't map to one entry in the gtf file 1-to-1 I guess."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b1d0ee6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shape: (212_986, 5)\n",
      "┌───────┬───────────┬───────────┬────────┬───────┐\n",
      "│ chrom ┆ start     ┆ end       ┆ strand ┆ reads │\n",
      "│ ---   ┆ ---       ┆ ---       ┆ ---    ┆ ---   │\n",
      "│ str   ┆ i64       ┆ i64       ┆ str    ┆ i64   │\n",
      "╞═══════╪═══════════╪═══════════╪════════╪═══════╡\n",
      "│ chr1  ┆ 944550    ┆ 945041    ┆ -      ┆ 1     │\n",
      "│ chr1  ┆ 944750    ┆ 945056    ┆ -      ┆ 1     │\n",
      "│ chr1  ┆ 944800    ┆ 945041    ┆ -      ┆ 141   │\n",
      "│ chr1  ┆ 944800    ┆ 945056    ┆ -      ┆ 284   │\n",
      "│ chr1  ┆ 944819    ┆ 945041    ┆ -      ┆ 1     │\n",
      "│ …     ┆ …         ┆ …         ┆ …      ┆ …     │\n",
      "│ chrX  ┆ 155898249 ┆ 155900496 ┆ +      ┆ 84    │\n",
      "│ chrX  ┆ 155900587 ┆ 155939700 ┆ +      ┆ 2     │\n",
      "│ chrX  ┆ 155900587 ┆ 155919812 ┆ +      ┆ 83    │\n",
      "│ chrX  ┆ 155919880 ┆ 155939700 ┆ +      ┆ 84    │\n",
      "│ chrX  ┆ 155939793 ┆ 155941882 ┆ +      ┆ 86    │\n",
      "└───────┴───────────┴───────────┴────────┴───────┘\n"
     ]
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import polars as pl\n",
    "import numpy as np\n",
    "\n",
    "reads = np.empty(0, dtype=np.float64)\n",
    "for file_path in input.all_introns_files:\n",
    "    df = pl.read_csv(file_path, separator='\\t')\n",
    "    reads = np.append(reads, df['reads'].to_numpy())\n",
    "\n",
    "plt.style.use('cowplot')\n",
    "fig, axes = plt.subplots(figsize=(5, 5))\n",
    "fig.suptitle('Intron read distribution')\n",
    "axes.hist(reads, bins=1000, alpha=0.5)\n",
    "axes.set_xlabel('Number of Reads')\n",
    "axes.set_ylabel('Count')\n",
    "ymin, ymax = axes.get_ylim()\n",
    "padding = 0.01 * ymax\n",
    "new_bottom = ymin - padding\n",
    "axes.set_ylim(bottom=new_bottom)\n",
    "\n",
    "axes.margins(y=0.01)\n",
    "plt.savefig(output.plot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a72b945",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
